### ISSUE TITLE
GPT2 Chatbot

### ISSUE THREAD

**Issue by ZeroMaxinumXZ** [minimaxir/gpt-2-simple#109]
Hi, I'm new to this stuff, and I'm trying to make a chatbot out of gpt2 using finetune... My question is, can I make gpt remember stuff like this:

Q: Hey, I'm Joe.
A: Hey, Joe.
Q: What's my name?
A:Joe.

Right now, it can barely remember anything I write to it. Do I just need a better more dialogue-focused dataset to finetune it on or is there something else I can do to make it remember? I'm using the colab notebook.

Thanks, I'm rather new to ML, so...

**Comment by ricknichols84** [comment]
Try this: https://github.com/huggingface/transfer-learning-conv-ai

**Comment by ZeroMaxinumXZ** [comment]
@ricknichols84, I would like to use this library if possible as it's more simple and I don't want to waste more computational resources.

 I have at least 300 question answer pairs in my dataset. Each question answer pairs is like

Q: What's your name?
A: Joe.
Q:  Hi...
A: Hey.

Do I need more data, or should I retrain on the bot's outputs and my inputs?

**Comment by r3nderer** [comment]
One issue you're going to run into is that content generation takes quite a while on GPT-2 models, so a real time chat bot might be somewhat difficult to obtain.

**Comment by johndpope** [comment]
@ZeroMaxinumXZ that link - https://medium.com/huggingface/how-to-build-a-state-of-the-art-conversational-ai-with-transfer-learning-2d818ac26313 sums things up in terms of architecture. Consider the architecture needs to recall / remember the conversation. This repo doesn't accommodate that. You can train and train but - you're going to get random responses on each call. Google has released pre-trained BERT models (only saw a 345mb) so you can avoid retraining things.  
Q) what are pros / cons of using BERT vs GTP2?
I don't have answer to that. 
Q) could you fine tune a BERT model as per gp2-simple?
I don't have answer to that either. 

I do know I've had great results with this repo and throwing it a terminator 2 and asking it to regenerate things. https://gist.github.com/johndpope/bc24f4f0ef186277aa6cde85a93cbc8a

**Comment by ZeroMaxinumXZ** [comment]
@r3nderer, I'm not looking for a real-time chatbot, just a quite-good one. Thanks for the help, anyway.

@johndpope, I created a 'memory-context bank" to hold the responses and queries, and fed that in as the prefix. It works out pretty good, and can hold a conversation for quite a while without me referencing other things in the conversation.

**Comment by johndpope** [comment]
@ZeroMaxinumXZ  - perhaps you can share a gist for others?

**Comment by nkrasner** [comment]
I didn't do anything extra with ML in mine, just the huggingface gpt2-xl model and some code I found which essentially made a talktotransformer.com type program. I store the generated/input conversation as it progresses and rerun it through the model. It fails sometimes (which could be fixed with fine-tuning) but when it succeeds, it is incredible the kinds of conversations it can have. Here's a link to a python notebook I made for it: https://colab.research.google.com/drive/1qbVsxvCjfTdWCqBzciLo1YeY0FhzQ-pn
